### DISCUSSION TITLE
WinPython & the Control Panel

### DISCUSSION THREAD

**Discussion by fizzixx** [winpython/winpython#1006] [Q&A] [answered]
Hello...

I'm trying out WinPython (Winpython64-3.9.5.0.exe) on a Windows 10 machine.

I think everything is working ok, except the Control Panel. There is no response double-clicking it. Everything else works - Spyder, Jupyter, JupyterLab, etc, but the Control Panel does nothing, no error message, nothing at all.

I've read many of the posts, and frankly I don't understand the lingo very well. I've done a lot of reading and searching online, but nothing helps. Seems like a lot of the documentation I have found is circular, and it often brings me back to where I started, and I'm out of ideas.

Is there a fix for this?

Thanks

**Comment by stonebig** [comment]
hi, does teh qtconsole shortcut work ? 
![image](https://user-images.githubusercontent.com/4312421/129392760-56ac2c51-893f-4d65-9a62-443a49924902.png)

If not, you may have installed "Winpython64-3.9.5.0dot.exe" and that doesn't include PyQt5, and so the control panel doesn't work. https://github.com/winpython/winpython/issues/834

**Reply by fizzixx** [reply]
Yes, it's installed and working.

**Reply by stonebig** [reply]
it also takes about 30" to start, on my PC

**Reply by fizzixx** [reply]
I wondered about that. I've double-clicked it and let it sit. Left it be for much longer than that, and nothing. Seems so odd that everything else works ok except this.

**Reply by stonebig** [reply]
hum, well. maybe:
- go to /scripts/wpcp.bat
- replace 
```
@echo off
call "%~dp0env_for_icons.bat" %*
cd/D "%WINPYWORKDIR1%"
"%WINPYDIR%\python.exe" -m winpython.controlpanel %*
```

per

```
@echo off
call "%~dp0env_for_icons.bat"
rem cd/D "%WINPYWORKDIR1%"
"%WINPYDIR%\python.exe" -m winpython.controlpanel 

```

- try again the icon

unless you did put winpython in a curious directory, like for https://github.com/winpython/winpython/issues/999, I may not have a clue.

**Reply by fizzixx** [reply]
stone....you're great, honestly....I don't know what "per" means. I don't understand what most of the 'lingo' means. It's waaaay too much effort to get a download to run properly, in my opinion. I'm not interested in software that requires more effort to install it and get it to run, than to actually use it. I just wanted to use it, but I guess that's not going to happen.
I GREATLY appreciate your help and willingness to help, but downloading a piece of software shouldn't have to require 'tech-support' to use.

Why don't you put this software together and post it so people can use it. You'd be a hero....even though you're a hero, to me, for trying so hard to help. thank you, but I'm uninstalling all of this and going back to square-one I guess.

**Reply by michaelkraft1512** [reply]
I have the same issue. I tried to change wpcp.bat as suggested by stonebig, but this does not solve the problem, WinPython Control Panel does not start. Same with VS Code. IPython Qt Console.exe DOES start.

**Reply by StephPer** [reply]
Hello, 

I have the same issue with WinPython 3.9.10 x64 on Windows Server 2016. I started WinPython Control Panel several times but did not get any reaction. I've checked the requirements and upgraded the Visual C++ redistributable package without success. I found an old post where someone describes to run the script controlpanel.py:

.\WP64-39100\python-3.9.10.amd64\Lib\site-package\winpython\controlpanel.py 

I got the following error messages:
`D:\Program files\WinPython\WPy64-39100\python-3.9.10.amd64\Lib\site-packages\winpython>python controlpanel.py
D:\Program files\WinPython\WPy64-39100\python-3.9.10.amd64\python.exe -c "import sys; print(sys.maxsize > 2**32)"  *  D:\Program files\WinPython\WPy64-39100\python-3.9.10.amd64
Traceback (most recent call last):
  File "D:\Program files\WinPython\WPy64-39100\python-3.9.10.amd64\Lib\site-packages\winpython\controlpanel.py", line 1014, in <module>
    main()
  File "D:\Program files\WinPython\WPy64-39100\python-3.9.10.amd64\Lib\site-packages\winpython\controlpanel.py", line 999, in main
    win = PMWindow()
  File "D:\Program files\WinPython\WPy64-39100\python-3.9.10.amd64\Lib\site-packages\winpython\controlpanel.py", line 494, in __init__
    self.setup_window()
  File "D:\Program files\WinPython\WPy64-39100\python-3.9.10.amd64\Lib\site-packages\winpython\controlpanel.py", line 544, in setup_window
    self.distribution_changed(sys.prefix)
  File "D:\Program files\WinPython\WPy64-39100\python-3.9.10.amd64\Lib\site-packages\winpython\controlpanel.py", line 828, in distribution_changed
    dist = wppm.Distribution(to_text_string(path))
  File "D:\Program files\WinPython\WPy64-39100\python-3.9.10.amd64\lib\site-packages\winpython\wppm.py", line 273, in __init__
    self.version, self.architecture = utils.get_python_infos(
  File "D:\Program files\WinPython\WPy64-39100\python-3.9.10.amd64\lib\site-packages\winpython\utils.py", line 387, in get_python_infos
    is_64 = python_query(
  File "D:\Program files\WinPython\WPy64-39100\python-3.9.10.amd64\lib\site-packages\winpython\utils.py", line 375, in python_query
    return exec_shell_cmd('%s -c "%s"' % (the_exe, cmd), path).splitlines()[0]
IndexError: list index out of range

D:\Program files\WinPython\WPy64-39100\python-3.9.10.amd64\Lib\site-packages\winpython>`

For testing, I've installed WinPython 3.9.4 x64 and here the WinPython Control Panel starts without any issue.

Any idea?

Kind regards
Stephan

**Reply by stonebig** [reply]
I've no clue why things happen badly on Windows Server 2016, I have none.
also:
- the "winpython control panel" was created for a time before "pip", 
- now that pip is quite functionnal and all packages are mostly wheel, you may just use pip,

... that said, if you are on multi user context and you don't move winpython, maybe there is this "shot-in-the-dark" you may try: 
    - \script\make_winpython_fix.bat

... to reverse it if no success,
   - copy your backup
   - or try \script\make_winpython_movable.bat

**Reply by StephPer** [reply]
Thank you for you response.

Actually, I use WinPyton Control Panel only do do the file associations and to add the shell extensions. My server has no internet connection so I don't plan to use the control panel to manage the packages. Is there another easy way to associate .py with Python, add the start menu entries and to get the Shell extensions?

Kind regards,
Stephan

**Reply by StephPer** [reply]
... mh, I tried the the suggested \scripts\make_winpython_fix.bat and once again it returned an error message:

```
patch pip and current launchers for non-move
D:\Program files\WinPython\WPy64-39100\python-3.9.10.amd64\python.exe -c "import sys; print(sys.maxsize > 2**32)"  *  D:\Program files\WinPython\WPy64-39100\python-3.9.10.amd64
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "D:\Program files\WinPython\WPy64-39100\python-3.9.10.amd64\lib\site-packages\winpython\wppm.py", line 273, in __init__
    self.version, self.architecture = utils.get_python_infos(
  File "D:\Program files\WinPython\WPy64-39100\python-3.9.10.amd64\lib\site-packages\winpython\utils.py", line 387, in get_python_infos
    is_64 = python_query(
  File "D:\Program files\WinPython\WPy64-39100\python-3.9.10.amd64\lib\site-packages\winpython\utils.py", line 375, in python_query
    return exec_shell_cmd('%s -c "%s"' % (the_exe, cmd), path).splitlines()[0]
IndexError: list index out of range
Press any key to continue . . .

```

Kind regards,
Stephan

**Reply by StephPer** [reply]
I'm a Python beginner but I recognized the issue occurs in the file '`utils.py`' and Python cannot split what `exec_shell_cmd` returned. 

The function `exec_shell_cmd` uses `subprocess.open` to execute the command

`python -c "import sys; print(sys.maxsize > 2**32)"`

This works well when I start it from the Python command prompt. Here it returns _True_. The function in the script returns null.

**Reply by StephPer** [reply]
Got it! WinPython 3.9.10 seems to have an issue with the space in the path. I've installed it under 'D:\Program Files'. I've isolated the function calling subprocess.popen and made several tests. At the end, I simply copied the entire WinPython folder to D:\WPy64-39100 and here I can start the WinPython Control Panel.

**Reply by StephPer** [reply]
To fix it, I changed line 375 in utils.py from

`    return exec_shell_cmd('%s -c "%s"' % (the_exe, cmd), path).splitlines()[0]`

to

`    return exec_shell_cmd('"%s" -c "%s"' % (the_exe, cmd), path).splitlines()[0]`

and now it works.

**Comment by stonebig** [comment]
there was also said at a moment (before pandemic I think) that "python instances" would do a graphical layer over pip.

**Comment by stonebig** [comment, marked as answer]
fixed in https://github.com/winpython/winpython/pull/1062
